test/ecto/multi_test.exs

```diff
@@ -330,4 +330,57 @@ defmodule Ecto.MultiTest do
       Multi.new |> Multi.run(:run, fun) |> Multi.run(:run, fun)
     end
   end
+
+  test "merge successfully flattens results" do
+    changeset = Changeset.change(%Comment{})
+    multi =
+      Multi.new
+      |> Multi.insert(:insert, changeset)
+      |> Multi.merge(fn data, multi ->
+        multi |> Multi.update(:update, Changeset.change(data.insert, x: 1))
+      end)
+
+    assert {:ok, data} = TestRepo.transaction(multi)
+    assert %Comment{} = data.insert
+    assert %Comment{} = data.update
+  end
+
+  test "merge rollbacks on errors" do
+    error = fn _ -> {:error, :error} end
+    ok    = fn _ -> {:ok, :ok} end
+
+    multi =
+      Multi.new
+      |> Multi.run(:outside_ok, ok)
+      |> Multi.merge(fn _, multi ->
+        multi
+        |> Multi.run(:inside_ok, ok)
+        |> Multi.run(:inside_error, error)
+      end)
+
+    assert {:error, :inside_error, :error, data} = TestRepo.transaction(multi)
+    assert :ok == data.outside_ok
+    assert :ok == data.inside_ok
+  end
+
+  test "merge does not allow repeated operations" do
+    fun = fn _ -> {:ok, :ok} end
+
+    Multi.new
+    |> Multi.merge(fn _, multi ->
+      assert_raise RuntimeError, ~r":run is already a member", fn ->
+        Multi.run(multi, :run, fun)
+      end
+    end)
+    |> Multi.run(:run, fun)
+
+
+    Multi.new
+    |> Multi.merge(fn _, multi -> Multi.run(multi, :run, fun) end)
```

Comment:
I think the approach where we don't pass a multi would be less confusing
and easier to compose (you can use any function that returns a multi
without having to pass yours).

On Sunday, June 12, 2016, Michał Muskała <notifications@github.com> wrote:

> In test/ecto/multi_test.exs
> <https://github.com/elixir-lang/ecto/pull/1482#discussion_r66727275>:
>
> > +  end
> > +
> > +  test "merge does not allow repeated operations" do
> > +    fun = fn _ -> {:ok, :ok} end
> > +
> > +    Multi.new
> > +    |> Multi.merge(fn _, multi ->
> > +      assert_raise RuntimeError, ~r":run is already a member", fn ->
> > +        Multi.run(multi, :run, fun)
> > +      end
> > +    end)
> > +    |> Multi.run(:run, fun)
> > +
> > +
> > +    Multi.new
> > +    |> Multi.merge(fn _, multi -> Multi.run(multi, :run, fun) end)
>
> I was wondering about that as well, what would be a better approach. I'm
> open to changing it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/ecto/pull/1482/files/ace09344c55e5c4227c379f4fe5140544a43a6bb#r66727275>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbtiTFzAUn_y7WzzBNb8ZyFm7WwHFks5qLHOmgaJpZM4Iz4Hm>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*